linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/http/AddForwardedHeaderConfig.scala

```diff
@@ -0,0 +1,121 @@
+
+@JsonTypeInfo(
+  use = JsonTypeInfo.Id.NAME,
+  include = JsonTypeInfo.As.PROPERTY,
+  property = "kind"
+)
+@JsonSubTypes(Array(
+  new JsonSubTypes.Type(value = classOf[IpLabelerConfig], name = "ip"),
+  new JsonSubTypes.Type(value = classOf[IpPortLabelerConfig], name = "ip:port"),
+  new JsonSubTypes.Type(value = classOf[RandomLabelerConfig], name = "random"),
+  new JsonSubTypes.Type(value = classOf[RouterLabelerConfig], name = "router"),
+  new JsonSubTypes.Type(value = classOf[StaticLabelerConfig], name = "static")
+))
+trait LabelerConfig {
+  @JsonIgnore
+  def mk(params: Stack.Params): AddForwardedHeader.Labeler
+}
+
+/** Labels an endpoint by its IP address, if it's known. */
+class IpLabelerConfig extends LabelerConfig {
+
+  @JsonIgnore
+  override def mk(params: Stack.Params) = AddForwardedHeader.Labeler.ClearIp
+}
+
+/** Labels an endpoint by its IP and port like `ip:port` */
+class IpPortLabelerConfig extends LabelerConfig {
+
+  @JsonIgnore
+  override def mk(params: Stack.Params) = AddForwardedHeader.Labeler.ClearIpPort
+}
+
+/** Generates a random obfuscated label. */
+class RandomLabelerConfig extends LabelerConfig {
+
+  @JsonIgnore
+  override def mk(params: Stack.Params) = AddForwardedHeader.Labeler.ObfuscatedRandom()
+}
+
+/** Uses the router name as an obfuscated label. */
+class RouterLabelerConfig extends LabelerConfig {
+
+  @JsonIgnore
+  override def mk(params: Stack.Params) =
+    AddForwardedHeader.Labeler.ObfuscatedStatic(params[Server.RouterLabel].label)
+}
+
+/** Uses the given string as an obfuscated label. */
+case class StaticLabelerConfig(label: String) extends LabelerConfig {
+
+  @JsonIgnore
+  override def mk(params: Stack.Params) =
+    AddForwardedHeader.Labeler.ObfuscatedStatic(label)
+}
+
+case class AddForwardedHeaderConfig(
+  by: Option[LabelerConfig],
+  `for`: Option[LabelerConfig]
+) {
+
+  /** Appends AddForwardedHeader params to the given params. */
+  @JsonIgnore
+  def ++:(params: Stack.Params): Stack.Params =
+    params + AddForwardedHeader.Enabled(true) + byParam(params) + forParam(params)
+
+  @JsonIgnore
+  private[this] def byParam(params: Stack.Params): AddForwardedHeader.Labeler.By =
+    by match {
+      case None => AddForwardedHeader.Labeler.By.default
+      case Some(config) => AddForwardedHeader.Labeler.By(config.mk(params))
+    }
+
+  @JsonIgnore
+  private[this] def forParam(params: Stack.Params): AddForwardedHeader.Labeler.For =
+    `for` match {
+      case None => AddForwardedHeader.Labeler.For.default
+      case Some(config) => AddForwardedHeader.Labeler.For(config.mk(params))
+    }
+}
+
+/**
+ * Translates AddForwardedHeaderConfig.Param into AddForwardedHeader
+ * configuration parameters.
+ */
+object AddForwardedHeaderConfig {
+
+  case class Param(config: Option[AddForwardedHeaderConfig])
+  implicit object Param extends Stack.Param[Param] {
+    val default = Param(None)
+  }
+
+  val module: Stackable[ServiceFactory[Request, Response]] =
```

Comment:
I wrote a big long comment about how I didn't understand what this was for, and in the process of writing that I figured it out.  (the values for forwarded for and forwarded by need to be computed here since they potentially depend on linkerd related values such as router label.  those values can then be passed down to the http router filter module that actually appends the headers)

So anyway, this is probably non-obvious enough to need some comments.